ktor-client/ktor-client-core/common/src/io/ktor/client/features/HttpPlainText.kt

```diff
@@ -13,51 +13,165 @@ import kotlinx.io.core.*
 
 /**
  * [HttpClient] feature that encodes [String] request bodies to [TextContent]
- * using a specific charset defined at [HttpPlainText.defaultCharset].
- * And also processes the response body as [String].
- * @param defaultCharset: default charset to use. [Charsets.UTF_8] by default.
+ * and processes the response body as [String].
  *
- * NOTE: the [HttpPlainText.defaultCharset] is the default one for your JVM so can change between servers!
- *       So please, specify one if you want consistent results in all your deployments.
+ * To configure charsets set following properties in [HttpPlainText.Config].
  */
-class HttpPlainText(var defaultCharset: Charset) {
+class HttpPlainText(
+    acceptCharsets: Set<Charset>,
+    charsetQuality: Map<Charset, Float>,
+    sendCharset: Charset?,
+    internal val responseCharsetFallback: Charset
+) {
+    private val requestCharset: Charset
+    private val acceptCharsetHeader: String
 
-    internal fun read(call: HttpClientCall, body: Input): String {
-        val actualCharset = call.response.charset() ?: call.responseConfig.defaultCharset
-        return body.readText(charset = actualCharset)
+    init {
+        val withQuality = charsetQuality.toList().sortedByDescending { it.second }
+        val withoutQuality = acceptCharsets.filter { !charsetQuality.containsKey(it) }.sortedBy { it.name }
+
+        acceptCharsetHeader = buildString {
+            withoutQuality.forEach {
+                if (length > 0) append(",")
+                append(it.name)
+            }
+
+            withQuality.forEach { (charset, quality) ->
+                if (length > 0) append(",")
+
+                check(quality in 0.0..1.0)
+                append("${charset.name};q=$quality")
+            }
+
+            if (isEmpty()) {
```

Comment:
Shouldn't we always accept `responseCharsetFallback`?

Reply:
On the contrary, always adding it to the list with no quality specified could make it always win (q = 1 by default)